static/app/stores/Commiters/CommittersContext.tsx

```diff
@@ -0,0 +1,274 @@
+        projectSlug: action.payload.projectSlug,
+        eventId: action.payload.eventId,
+      });
+      return {
+        ...state,
+        [key]: {
+          committers: [],
+          committersLoading: true,
+          committersError: undefined,
+        },
+      };
+    }
+    case 'set error': {
+      const key = makeCommitterStoreKey({
+        organizationSlug: action.payload.organizationSlug,
+        projectSlug: action.payload.projectSlug,
+        eventId: action.payload.eventId,
+      });
+
+      return {
+        ...state,
+        [key]: {
+          committers: [],
+          committersLoading: false,
+          committersError: true,
+        },
+      };
+    }
+    case 'reset': {
+      return {};
+    }
+    default: {
+      exhaustive(action);
+      return state;
+    }
+  }
+}
+
+export const CommittersContext = createContext<
+  [CommittersState, React.Dispatch<CommittersAction>] | null
+>(null);
+
+interface CommittersContextProviderProps {
+  children: React.ReactNode;
+  initialState?: CommittersState;
+}
+
+export function CommittersProvider(props: CommittersContextProviderProps) {
+  const contextValue = useReducer(committersReducer, props.initialState ?? {});
+
+  return (
+    <CommittersContext.Provider value={contextValue}>
+      {props.children}
+    </CommittersContext.Provider>
+  );
+}
+
+export function useCommitters(): [CommittersState, React.Dispatch<CommittersAction>] {
+  const context = useContext(CommittersContext);
+
+  if (!context) {
+    throw new Error('useCommitters called outside of CommittersContext.Provider');
+  }
+
+  return context;
+}
+
+interface RequiredWithCommittersProps {
+  event: Event;
+  organization: Organization;
+  project: Project | AvatarProject;
+  group?: Group;
+}
+
+export interface WithCommittersProps {
+  committers: Committer[] | undefined;
+}
+
+export function withCommitters<P extends WithCommittersProps>(
+  Component: React.ComponentType<P>
+) {
+  const wrappedComponent: React.ComponentType<
+    Omit<P, keyof WithCommittersProps> & RequiredWithCommittersProps
+  > = (props): React.ReactElement => {
+    const api = useApi();
+    const [state, dispatch] = useCommitters();
+
+    const key = useMemo(() => {
+      return makeCommitterStoreKey({
+        organizationSlug: props.organization.slug,
+        projectSlug: props.project.slug,
+        eventId: props.event.id,
+      });
+    }, [props.organization.slug, props.project.slug, props.event.id]);
+
+    useEffect(() => {
+      if (!props.group?.firstRelease) {
+        return undefined;
+      }
+
+      if (state[key]?.committers || state[key]?.committersLoading) {
+        return undefined;
+      }
+
+      dispatch({
+        type: 'start loading',
+        payload: {
+          eventId: props.event.id,
+          organizationSlug: props.organization.slug,
+          projectSlug: props.project.slug,
+        },
+      });
+
+      let unmounted = false;
```

Comment:
Yeah, we dont have any guidelines for this, I've used the let pattern in a few other places, so I just went with this. I'm down to use whatever is most common in other places.

Reply:
yeah, looking at it more closely, this isn't really `mounted` though, it's actually just the loaded scope within this `useEffect`, which will stop the state being set after the promise is finished if another useEffect was called because of the cleanup, it's a bit misleading, imo. It's actually be more accurate to have it as `requestCancelled` since that is it's effect.

Reply:
@k-fish good point, let me rename this!